### DISCUSSION TITLE
Global connections

### DISCUSSION THREAD

**Discussion by Janpot** [mui/toolpad#1376] [Early feedback]
## Motivation

* https://github.com/mui/mui-toolpad/issues/1065
* Give users the ability to manage all their connections to datasources centrally, and share connections across applications
* Application spec is a shareable artifact. Connections contain sensitive information. The two concepts are hard to reconcile
* Give the application editor more screen real estate to focus better on app building
* When we want to unlock templates, or application sharing, we will need to be able to pass around an application spec without connections embedded, and a way to link a connection upon instantiation.

## Ideal goals

* Connections moved from app to global.
* True handling of secrets:
  * stored encrypted (ok now)
  * never transferred server -> browser, only browser -> server (not ok now)
  * not part of shareable artifacts (not ok now)
* connections can be created/managed from the applications overview page
* connections can be created from an application, without leaving the application
* Unlock templates, and application sharing

## Design

Connections overview
![image](https://user-images.githubusercontent.com/2109932/203563788-88e17ce4-1659-4ac0-9944-f68183801b09.png)

@gerdadesign

**Comment by Janpot** [comment]
### Topic: list view

More general as it applies to apps as well:
* What's the added value of the grid view? 
* Would there be benefit in moving the list to a DataGrid instead along the lines of https://mui.com/x/react-data-grid/demo/

**Reply by oliviertassinari** [reply]
> What's the added value of the grid view?

Maybe if we had screenshots like in Figma or Framer I would understand it

<img width="376" alt="Screenshot 2022-11-23 at 15 21 27" src="https://user-images.githubusercontent.com/3165635/203570119-a0986cab-332e-419c-9a93-247449cfbdf2.png">

but we don't, so 👍

**Reply by Janpot** [reply]
Implementing screenshots properly will require rendering pages on the server. i.e. in a headless browser. This will bloat the already hefty docker image for self-hosters. apps like figma don't have to self host, they can even allocate a separate service for this. Wouldn't recommend for this for now. 

Alternatively one could think about a solution based on `html2canvas`. We could try rendering it in an iframe while deploying. But it wouldn't be without cost:
* It bloats user experience, will take a lot more time to do a deploy. It could take a long time before the page is rendered, we'd be blocking the deploy on that. This depends on user input, it's unpredictable.
* Not all web features are supported by this library. It doesn't seem that actively maintained.
* which page to capture?

**Reply by gerdadesign** [reply]
The question about Grid View was also something I was exploring in Figma. The main benefit of this does seem to be including a screenshot, which seems like a valuable thing to have when looking at all your apps at a glance. It would help to quickly visually choose which one you are looking for. 

The below shows the sections we discussed having on the homepage in both Grid View and List View. An image makes most sense for some; list for others. 

However, if adding a preview image adds a lot of extra bloat and complication, we could simply just use the list view for now? 

| Apps | Global Connections | Global Components |
| --- | --- | --- | 
| Grid: app preview | Grid: not sure if an image makes sense here (unless we used that space for the img of the connection type, but I think the secondary info + icon is a better place for that) | Grid: component preview image | 
| <img width="635" alt="Component grid view" src="https://user-images.githubusercontent.com/106991024/203600994-b31861c3-1bc6-4da7-99ae-c44f3dc875c4.png"> | <img width="607" alt="Connection grid view" src="https://user-images.githubusercontent.com/106991024/203600996-883e9166-1b90-4b0e-adcd-db2aabad8afc.png"> | <img width="744" alt="Apps grid view" src="https://user-images.githubusercontent.com/106991024/203600997-3bd6573f-77ef-4322-8f17-aee6971c7f7a.png"> | 
| List | List | List: image might be most helpful here | 
| <img width="777" alt="App list view" src="https://user-images.githubusercontent.com/106991024/203603660-b4483af6-49c4-482c-a097-f9162beeb166.png"> | <img width="776" alt="Connection list view" src="https://user-images.githubusercontent.com/106991024/203603658-0c91c8c5-5331-4c10-a26f-89d5cbf1e8f6.png"> | <img width="774" alt="Component list view" src="https://user-images.githubusercontent.com/106991024/203603653-632c4e3b-62d2-48e3-a395-abd716c9ba7a.png"> |

**Reply by oliviertassinari** [reply]
I have created #1399 to keep track of this idea. I agree that it's for later. It would be better to only build this once it's requested by enough people. It's not core to the value prop of the product.

**Reply by prakhargupta1** [reply]
Thanks Olivier, for creating the issue. In our last [meeting](https://www.notion.so/mui-org/Planning-W48-49-cc9a274727c9463bac08a9e5111c0fad#2a08ccbc951a4c068602a402fa882b3a) we discussed doing the card-view of Connections later. For apps, we can keep what we have currently and improve our List view. 

Having apps in a searchable, well-designed table should solve the pain you've mentioned in #1399

**Reply by prakhargupta1** [reply]
An example of List-view, card-view I found was Codesandbox. Looks like their list view is a table.
<img width="1431" alt="Screenshot 2022-12-02 at 10 08 31 AM" src="https://user-images.githubusercontent.com/92228082/205216343-9d1b8daf-0e32-466f-9dbd-552e8dae136c.png">

**Comment by WangLarry** [comment]
A new Retool-like open source Low-code/No-code platforms: https://github.com/openblocks-dev/openblocks
It has workspace feature, have All Apps,Modules,Trash,Folders,Query Library,Data Sources etc in workspace level.

@Janpot

**Reply by Janpot** [reply]
Thanks for sharing. openblocks is on our radar. Well thought out trade-offs made in terms of simplicity and organization

**Reply by WangLarry** [reply]
typebot.io has also workspace feature, https://github.com/baptisteArno/typebot.io
It's code struct is clean and modern.
BTW: it's 'models' is interesting, which use zod and prisma schema to define data model.

**Reply by oliviertassinari** [reply]
@WangLarry Thanks for sharing. It has definitely been on our radar.

I'm curious, what did you find the most interesting in today's Toolpad and its potential?

Regarding where we are heading. The opportunity/pain point/niche in the space that we are exploring so far is to deeply integrate a Retool-like product experience with the benefits of pro-code. It's about allowing developers to stay in sync with their backend code (database schema or APIs), their design systems (e.g. Material UI custom themes), and the frontend pages of the company(e.g. Next.js). I think that our ideal users would be ones using https://refine.dev/, https://marmelab.com/react-admin/, https://github.com/PanJiaChen/vue-element-admin but still not willing to jump low-code with say, Appsmith but that are frustrated when using these tools.

**Comment by prakhargupta1** [comment]
This is about the old Toolpad. I am closing it.
